`__slots__` requires explicit type definition and must be `Any` (and nothing else) when used with `Protocol`s

``` python
# test_case1.py
from __future__ import annotations
from typing import Protocol, runtime_checkable

@runtime_checkable
class Foo(Protocol):
    __slots__ = ()
    def foo(self):
        ...

class Bar:
    __slots__ = ("_t",)
    def __init__(self):
        self._t = True
    def foo(self):
        pass

print(f"isinstance(Bar(), Foo) == {isinstance(Bar(), Foo)}")  # <- prints … == True
foo: Foo = Bar()  # <- errors here
```

```
% python test_case1.py
isinstance(Bar(), Foo) == True
% mypy --version
mypy 0.910
% mypy --config-file=/dev/null test_case1.py
/dev/null: No [mypy] section in config file
test_case1.py:19: error: Incompatible types in assignment (expression has type "Bar", variable has type "Foo")
test_case1.py:19: note: Following member(s) of "Bar" have conflicts:
test_case1.py:19: note:     __slots__: expected "Tuple[]", got "Tuple[str]"
```

The "fix" is to provide `Tuple[str, ...]` as the type for `__slots__`:

``` python
# test_case2.py
from __future__ import annotations
from typing import Protocol, Tuple, runtime_checkable

@runtime_checkable
class Foo(Protocol):
    __slots__: Tuple[str, ...] = ()
    def foo(self):
        ...

class Bar:
    __slots__ : Tuple[str, ...] = ("_t",)
    def __init__(self):
        self._t = True
    def foo(self):
        pass

print(f"isinstance(Bar(), Foo) == {isinstance(Bar(), Foo)}")  # still True
foo: Foo = Bar()  # <- works now?
```

```
% mypy --config-file=/dev/null test_case2.py
/dev/null: No [mypy] section in config file
Success: no issues found in 1 source file
```

Shouldn't mypy already know the type of `__slots__`?

Here's another oddity _**not**_ fixed by typing `__slots___`:

``` python
# test_case3.py
from fractions import Fraction
from typing import Protocol, Tuple, TypeVar, runtime_checkable
_T_co = TypeVar("_T_co", covariant=True)

@runtime_checkable
class _RationalInitializerT(Protocol[_T_co]):
    __slots__: Tuple[str, ...] = ()
    def __call__(self, numerator: int, denominator: int) -> _T_co:
        ...

f: _RationalInitializerT = Fraction  # <- errors
```

```
% mypy --config-file=/dev/null test_case3.py
/dev/null: No [mypy] section in config file
test_case3.py:12: error: Incompatible types in assignment (expression has type "Type[Fraction]", variable has type "_RationalInitializerT[Any]")
Found 1 error in 1 file (checked 1 source file)
```

Now, remove `__slots__`:

``` python
# test_case4.py
from fractions import Fraction
from typing import Protocol, Tuple, TypeVar, runtime_checkable
_T_co = TypeVar("_T_co", covariant=True)

@runtime_checkable
class _RationalInitializerT(Protocol[_T_co]):
    # __slots__: Tuple[str, ...] = ()  # <- commented out
    def __call__(self, numerator: int, denominator: int) -> _T_co:
        ...

f: _RationalInitializerT = Fraction  # <- works now?
```

```
% mypy --config-file=/dev/null test_case4.py
/dev/null: No [mypy] section in config file
Success: no issues found in 1 source file
```

I found another case for which I cannot find a work-around, specifically where `__slots__` for the value's type doesn't precisely match `__slots__` for the protocol's type:

``` python
# test_case.py
from abc import abstractmethod
from typing import Any, Iterable, Protocol, Tuple, Union, runtime_checkable


class Foo:
    __slots__ = ("bar", "baz")

    def foo(self) -> None:
        pass


@runtime_checkable
class SupportsFoo(Protocol):
    __slots__: Union[
        str,
        Tuple[str],
        Tuple[str, str],
        Tuple[str, str, str],
        Tuple[str, ...],
        Iterable[str],
        Any,  # <-- adding this doesn't even help
    ] = ()

    @abstractmethod
    def foo(self) -> None:
        pass


@runtime_checkable
class SupportsFooExactMatchWorks(Protocol):
    __slots__: Tuple[str, str] = ()  # type: ignore

    @abstractmethod
    def foo(self) -> None:
        pass


@runtime_checkable
class SupportsFooAnyWorks(Protocol):
    __slots__: Any = ()  # type: ignore

    @abstractmethod
    def foo(self) -> None:
        pass


v1: SupportsFoo = Foo()  # <-- error
v2: SupportsFooExactMatchWorks = Foo()  # works: precise type match on __slots__
v3: SupportsFooAnyWorks = Foo()  # works: __slots__: Any, and nothing else
```

``` sh
% mypy --config-file=/dev/null test_case.py
/dev/null: No [mypy] section in config file
test_case.py:48: error: Incompatible types in assignment (expression has type "Foo", variable has type "SupportsFoo")
test_case.py:48: note: Following member(s) of "Foo" have conflicts:
test_case.py:48: note:     __slots__: expected "Union[str, Tuple[str], Tuple[str, str], Tuple[str, str, str], Tuple[str, ...], Iterable[str], Any]", got "Tuple[str, str]"
Found 1 error in 1 file (checked 1 source file)
```

What this effectively means is that if `Protocol`s define `__slots__` they must be explicitly typed as `Any` and nothing else. That seems wrong.

Collecting related issues:

#11821 (marked as fixed)
#11885 (still in draft, blocked on #11891)

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -2801,34 +2801,6 @@ class TempNode(Expression):
         return visitor.visit_temp_node(self)
 
 
-# Special attributes not collected as protocol members by Python 3.12
-# See typing.EXCLUDED_ATTRIBUTES and typing._get_protocol_attrs
-EXCLUDED_ATTRIBUTES = frozenset(
-    {
-        "__parameters__",
-        "__orig_bases__",
-        "__orig_class__",
-        "_is_protocol",
-        "_is_runtime_protocol",
-        "__protocol_attrs__",
-        "__callable_proto_members_only__",
-        "__type_params__",
-        "__abstractmethods__",
-        "__annotations__",
-        "__dict__",
-        "__doc__",
-        "__init__",
-        "__module__",
-        "__new__",
-        "__slots__",
-        "__subclasshook__",
-        "__weakref__",
-        "__class_getitem__",
-        "_MutableMapping__marker",
-    }
-)
-
-
 class TypeInfo(SymbolNode):
     """The type structure of a single class.
 
@@ -3045,6 +3017,24 @@ class TypeInfo(SymbolNode):
         "is_intersection",
     ]
 
+    # Special attributes not collected as protocol members by Python 3.12
+    # See typing._SPECIAL_NAMES
+    EXCLUDED_ATTRIBUTES: Final = frozenset(
+        {
+            "__abstractmethods__",
+            "__annotations__",
+            "__dict__",
+            "__doc__",
+            "__init__",
+            "__module__",
+            "__new__",
+            "__slots__",
+            "__subclasshook__",
+            "__weakref__",
+            "__class_getitem__",  # Since Python 3.9
+        }
+    )
+
     def __init__(self, names: SymbolTable, defn: ClassDef, module_name: str) -> None:
         """Initialize a TypeInfo."""
         super().__init__()
@@ -3143,7 +3133,7 @@ class TypeInfo(SymbolNode):
                     if isinstance(node.node, (TypeAlias, TypeVarExpr, MypyFile)):
                         # These are auxiliary definitions (and type aliases are prohibited).
                         continue
-                    if name in EXCLUDED_ATTRIBUTES:
+                    if name in self.EXCLUDED_ATTRIBUTES:
                         continue
                     members.add(name)
         return sorted(list(members))

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -2801,6 +2801,34 @@ class TempNode(Expression):
         return visitor.visit_temp_node(self)
 
 
+# Special attributes not collected as protocol members by Python 3.12
+# See typing.EXCLUDED_ATTRIBUTES and typing._get_protocol_attrs
+EXCLUDED_ATTRIBUTES = frozenset(
+    {
+        "__parameters__",
+        "__orig_bases__",
+        "__orig_class__",
+        "_is_protocol",
+        "_is_runtime_protocol",
+        "__protocol_attrs__",
+        "__callable_proto_members_only__",
+        "__type_params__",
+        "__abstractmethods__",
+        "__annotations__",
+        "__dict__",
+        "__doc__",
+        "__init__",
+        "__module__",
+        "__new__",
+        "__slots__",
+        "__subclasshook__",
+        "__weakref__",
+        "__class_getitem__",
+        "_MutableMapping__marker",
+    }
+)
+
+
 class TypeInfo(SymbolNode):
     """The type structure of a single class.
 
@@ -3114,6 +3142,8 @@ class TypeInfo(SymbolNode):
                 for name, node in base.names.items():
                     if isinstance(node.node, (TypeAlias, TypeVarExpr, MypyFile)):
                         # These are auxiliary definitions (and type aliases are prohibited).
+                        continue
+                    if name in EXCLUDED_ATTRIBUTES:
                         continue
                     members.add(name)
         return sorted(list(members))

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -2801,6 +2801,25 @@ class TempNode(Expression):
         return visitor.visit_temp_node(self)
 
 
+# Special attributes not collected as protocol members by Python 3.12
+# See typing._SPECIAL_NAMES
+EXCLUDED_PROTOCOL_ATTRIBUTES: Final = frozenset(
+    {
+        "__abstractmethods__",
+        "__annotations__",
+        "__dict__",
+        "__doc__",
+        "__init__",
+        "__module__",
+        "__new__",
+        "__slots__",
+        "__subclasshook__",
+        "__weakref__",
+        "__class_getitem__",  # Since Python 3.9
+    }
+)
+
+
 class TypeInfo(SymbolNode):
     """The type structure of a single class.
 
@@ -3017,24 +3036,6 @@ class TypeInfo(SymbolNode):
         "is_intersection",
     ]
 
-    # Special attributes not collected as protocol members by Python 3.12
-    # See typing._SPECIAL_NAMES
-    EXCLUDED_ATTRIBUTES: Final = frozenset(
-        {
-            "__abstractmethods__",
-            "__annotations__",
-            "__dict__",
-            "__doc__",
-            "__init__",
-            "__module__",
-            "__new__",
-            "__slots__",
-            "__subclasshook__",
-            "__weakref__",
-            "__class_getitem__",  # Since Python 3.9
-        }
-    )
-
     def __init__(self, names: SymbolTable, defn: ClassDef, module_name: str) -> None:
         """Initialize a TypeInfo."""
         super().__init__()
@@ -3133,7 +3134,7 @@ class TypeInfo(SymbolNode):
                     if isinstance(node.node, (TypeAlias, TypeVarExpr, MypyFile)):
                         # These are auxiliary definitions (and type aliases are prohibited).
                         continue
-                    if name in self.EXCLUDED_ATTRIBUTES:
+                    if name in EXCLUDED_PROTOCOL_ATTRIBUTES:
                         continue
                     members.add(name)
         return sorted(list(members))

--- a/test-data/unit/check-protocols.test
+++ b/test-data/unit/check-protocols.test
@@ -2789,6 +2789,42 @@ class A(Protocol):
 
 [builtins fixtures/tuple.pyi]
 
+[case testProtocolSlotsIsNotProtocolMember]
+# https://github.com/python/mypy/issues/11884
+from typing import Protocol
+
+class Foo(Protocol):
+    __slots__ = ()
+class NoSlots:
+    pass
+class EmptySlots:
+    __slots__ = ()
+class TupleSlots:
+    __slots__ = ('x', 'y')
+class StringSlots:
+    __slots__ = 'x y'
+def foo(f: Foo):
+    pass
+
+# All should pass:
+foo(NoSlots())
+foo(EmptySlots())
+foo(TupleSlots())
+foo(StringSlots())
+[builtins fixtures/tuple.pyi]
+
+[case testProtocolSlotsAndRuntimeCheckable]
+from typing import Protocol, runtime_checkable
+
+@runtime_checkable
+class Foo(Protocol):
+    __slots__ = ()
+class Bar:
+    pass
+issubclass(Bar, Foo)  # Used to be an error, when `__slots__` counted as a protocol member
+[builtins fixtures/isinstance.pyi]
+[typing fixtures/typing-full.pyi]
+
 [case testNoneVsProtocol]
 # mypy: strict-optional
 from typing_extensions import Protocol

--- a/test-data/unit/check-protocols.test
+++ b/test-data/unit/check-protocols.test
@@ -2825,6 +2825,29 @@ issubclass(Bar, Foo)  # Used to be an error, when `__slots__` counted as a proto
 [builtins fixtures/isinstance.pyi]
 [typing fixtures/typing-full.pyi]
 
+
+[case testProtocolWithClassGetItem]
+# https://github.com/python/mypy/issues/11886
+from typing import Any, Iterable, Protocol, Union
+
+class B:
+    ...
+
+class C:
+    def __class_getitem__(cls, __item: Any) -> Any:
+        ...
+
+class SupportsClassGetItem(Protocol):
+    __slots__: Union[str, Iterable[str]] = ()
+    def __class_getitem__(cls, __item: Any) -> Any:
+        ...
+
+b1: SupportsClassGetItem = B()
+c1: SupportsClassGetItem = C()
+[builtins fixtures/tuple.pyi]
+[typing fixtures/typing-full.pyi]
+
+
 [case testNoneVsProtocol]
 # mypy: strict-optional
 from typing_extensions import Protocol

--- a/test-data/unit/check-protocols.test
+++ b/test-data/unit/check-protocols.test
@@ -2803,6 +2803,10 @@ class TupleSlots:
     __slots__ = ('x', 'y')
 class StringSlots:
     __slots__ = 'x y'
+class InitSlots:
+    __slots__ = ('x',)
+    def __init__(self) -> None:
+        self.x = None
 def foo(f: Foo):
     pass
 
@@ -2811,6 +2815,7 @@ foo(NoSlots())
 foo(EmptySlots())
 foo(TupleSlots())
 foo(StringSlots())
+foo(InitSlots())
 [builtins fixtures/tuple.pyi]
 
 [case testProtocolSlotsAndRuntimeCheckable]

